File: trees_and_graphs/binary_trees/dfs/python3/path_sum_iii.py
from collections import defaultdict
from typing import DefaultDict, List, Optional, Tuple

from tree_node import TreeNode


class PathSum(object):
    """
    Finds the number of paths in a binary tree where the sum of node values equals a given target.

    Attributes
    ----------
    target : int
        The target sum to find in paths.
    count : int
        The number of valid paths found.
    curr_sum : int
        The running sum of the current path.
    curr_sum_counts : DefaultDict[int, int]
        A hashmap tracking prefix sums and their frequencies.
    """

    def __init__(self, target: int) -> None:
        """
        Initializes the PathSum instance.

        Parameters
        ----------
        target : int
            The target sum to find in paths.
        """
        self.target: int = target
        self.count: int = 0
        self.curr_sum: int = 0
        self.curr_sum_counts: DefaultDict[int, int] = defaultdict(int)

    def depth_first_search(self, node: Optional[TreeNode], depth: int) -> None:
        """
        Performs a depth-first search to count paths that sum to the target.

        Parameters
        ----------
        node : Optional[TreeNode]
            The current node being processed.
        depth : int
            The depth of the current node in the tree.
        """
        print_prefix: str = f"{'  ' * depth}Depth {depth} | "
        if not node:
            print(f"{print_prefix}Empty node")
            return None

        assert isinstance(node.data, int)
        print(f"{print_prefix}Current sum ({self.curr_sum}) + Node data ({node.data}) = ", end="")
        print(f"{self.curr_sum + node.data}")
        self.curr_sum += node.data

        print(f"{print_prefix}Current sum ({self.curr_sum}) == target ({self.target}) is ", end="")
        print(f"{self.curr_sum == self.target}")
        if self.curr_sum == self.target:
            print(f"{print_prefix}Total count ({self.count}) + 1 = {self.count + 1}")
            self.count += 1

        print(f"{print_prefix}Current sum ({self.curr_sum}) - target ({self.target}) = ", end="")
        print(f"{self.curr_sum - self.target} is in the counts map? ", end="")
        print(f"{(self.curr_sum - self.target) in self.curr_sum_counts}")
        if (self.curr_sum - self.target) in self.curr_sum_counts:
            print(f"{print_prefix}Total count ({self.count}) ", end="")
            print(f" + Counts for {self.curr_sum - self.target} ", end="")
            print(f"({self.curr_sum_counts[self.curr_sum - self.target]}) = ", end="")
            print(f"{self.count + self.curr_sum_counts[self.curr_sum - self.target]}")
            self.count += self.curr_sum_counts[self.curr_sum - self.target]

        print(f"{print_prefix}Counts hash map before = {self.curr_sum_counts}")
        self.curr_sum_counts[self.curr_sum] += 1
        print(f"{print_prefix}Counts hash map after = {self.curr_sum_counts}")

        self.depth_first_search(node.left, depth + 1)
        self.depth_first_search(node.right, depth + 1)

        backtrace_prefix: str = f"{print_prefix} Backtrace | "
        print(f"{backtrace_prefix}Count for {self.curr_sum} ", end="")
        print(f"({self.curr_sum_counts[self.curr_sum]}) - 1 = ", end="")
        print(f"{self.curr_sum_counts[self.curr_sum] - 1}")
        self.curr_sum_counts[self.curr_sum] -= 1

        print(f"{backtrace_prefix}Remaining count for {self.curr_sum} ", end="")
        print(f" = {self.curr_sum_counts[self.curr_sum]}")
        if not self.curr_sum_counts[self.curr_sum]:
            print(f"{backtrace_prefix}Remove key {self.curr_sum_counts[self.curr_sum]}")
            del self.curr_sum_counts[self.curr_sum]

        print(f"{backtrace_prefix}Current sum ({self.curr_sum}) - Node data ({node.data}) ", end="")
        print(f" = {self.curr_sum - node.data}")
        self.curr_sum -= node.data
        return None

    def search(self, root: Optional[TreeNode]) -> int:
        """
        Search for paths that sum to the target.

        Parameters
        ----------
        root : Optional[TreeNode]
            The root node of the binary tree.

        Returns
        -------
        int
            The number of valid paths found.
        """
        if not root:
            return 0
        self.depth_first_search(root, 0)
        return self.count


def main() -> int:
    test_cases: List[Tuple[List[Optional[int]], int]] = [
        ([5, 4, 8, 11, None, 13, 4, 7, 2, None, None, 5, 1], 22),
        ([0, 1, 1], 1),
        ([10, 5, -3, 3, 2, None, 11, 3, -2, None, 1], 5),
    ]
    for node_data, target in test_cases:
        print("Case " + "-" * 60, "\n")
        root: Optional[TreeNode] = TreeNode.construct_binary_tree(values=node_data)
        path_sum: PathSum = PathSum(target=target)
        num_paths: int = path_sum.search(root)
        print(f"\nNumber of searched paths = {num_paths}\n")
    return 0


if __name__ == "__main__":
    main()

File: trees_and_graphs/binary_trees/dfs/c++/path_sum_iii/path_sum_iii.cpp
// path_sum_iii.cpp
#include "path_sum_iii.hpp"
#include "tree_node.hpp"
#include <unordered_map>
#include <functional>
#include <variant>
#include <stdexcept>
#include <limits>
#include <cstdint>

int pathSum(TreeNode::ptr &root, int target)
{
    // Total paths found
    int count = 0;
    // If the tree is empty, return 0 path found
    if (!root)
        return count;

    // Define Sum type that can be easily changed between int and int64_t
    using Sum = int;

    // Current sum and counts hash map
    std::unordered_map<Sum, int> currSumCounts;
    Sum currSum = 0;

    // Helper function to detect potential integer overflow
    auto wouldOverflow = [](Sum a, int b) -> bool
    {
        if (b > 0 && a > std::numeric_limits<Sum>::max() - b)
            return true;
        if (b < 0 && a < std::numeric_limits<Sum>::min() - b)
            return true;
        return false;
    };

    // Depth first search lambda function
    std::function<void(TreeNode::ptr &)> depthFirstSearch = [&](TreeNode::ptr &node)
    {
        // Empty node
        if (!node)
            return;

        // Ensure node data is an integer
        if (!std::holds_alternative<int>(node->data))
            throw std::runtime_error("Non-integer data encountered in TreeNode");

        // Extract integer value and check for potential overflow before adding
        int currData = std::get<int>(node->data);

        // Check if adding current value would cause overflow
        if (wouldOverflow(currSum, currData))
        {
            // Switch to large number handling path
            handleLargePath(static_cast<std::int64_t>(currSum), node, count, target);
            return;
        }

        // Add current value to sum (safe now that we've checked for overflow)
        currSum += currData;

        // Check if the current sum matches the target, if so, we've found a path
        if (currSum == target)
            count++;

        // Check if (current sum - target) is already in the hash map
        Sum diff = currSum - target;
        if (auto searchResults = currSumCounts.find(diff); searchResults != currSumCounts.end())
        {
            // The count represents the number of valid paths, add them to the total count
            count += searchResults->second;
        }

        // Increment the count of the current sum
        currSumCounts[currSum]++;

        depthFirstSearch(node->left);
        depthFirstSearch(node->right);

        // Backtrace by decrementing the count of the current sum
        currSumCounts[currSum]--;

        // If the count of the current sum has reached zero, remove it
        if (!currSumCounts[currSum])
            currSumCounts.erase(currSum);
        // Subtract the current node data from the sum
        currSum -= currData;

        return;
    };

    depthFirstSearch(root);

    return count;
}

void handleLargePath(std::int64_t initialSum, TreeNode::ptr &node, int &count, const int target)
{
    // Switch to 64-bit arithmetic for paths with large numbers
    using BigSum = std::int64_t;
    std::unordered_map<BigSum, int> largeSumCounts;
    BigSum currSum = initialSum;
    BigSum bigTarget = static_cast<BigSum>(target);

    // Helper function to detect 64-bit overflow
    auto wouldOverflow64 = [](BigSum a, int b) -> bool
    {
        if (b > 0 && a > std::numeric_limits<BigSum>::max() - b)
            return true;
        if (b < 0 && a < std::numeric_limits<BigSum>::min() - b)
            return true;
        return false;
    };

    std::function<void(TreeNode::ptr &)> largeDFS = [&](TreeNode::ptr &node)
    {
        if (!node)
            return;

        if (!std::holds_alternative<int>(node->data))
            throw std::runtime_error("Non-integer data encountered in TreeNode");

        int currData = std::get<int>(node->data);

        // Check for overflow even in 64-bit path
        if (wouldOverflow64(currSum, currData))
        {
            throw std::overflow_error("Path sum too large even for 64-bit arithmetic");
        }

        currSum += static_cast<BigSum>(currData);

        if (currSum == bigTarget)
            count++;

        BigSum diff = currSum - bigTarget;
        if (auto searchResults = largeSumCounts.find(diff); searchResults != largeSumCounts.end())
        {
            count += searchResults->second;
        }

        largeSumCounts[currSum]++;

        largeDFS(node->left);
        largeDFS(node->right);

        largeSumCounts[currSum]--;
        if (!largeSumCounts[currSum])
            largeSumCounts.erase(currSum);
        currSum -= static_cast<BigSum>(currData);
    };

    // Start the large number DFS from the current node
    largeDFS(node);
}
